### PR TITLE
feat: enable http cache

### PR DESCRIPTION
**Pull request by sileht**

### CODE REVIEW

File: mergify_engine/__init__.py [hunk 1]
```diff
@@ -12,12 +12,16 @@
 # License for the specific language governing permissions and limitations
 # under the License.
 
+import cachecontrol
+from cachecontrol.caches import FileCache
```
Comment (jd) [inline mergify_engine/__init__.py]: I feel like this is going to be lost on restart right?

File: setup.cfg [hunk 2]
```diff
@@ -22,6 +22,8 @@ include_package_data = true
 
 install_requires =
     attrs
+    cachecontrol
+    lockfile
```
Comment (jd) [inline setup.cfg]: `cachecontrol[filecache]` has the right deps

Comment (sileht) [reply setup.cfg]: done

File: mergify_engine/__init__.py (hunk 1 above)
Comment (sileht) [reply mergify_engine/__init__.py]: It's was for my testing, it's going to be redis at this end, I have already prepared a new redis instance in production. We now have 5 servers, so we can't downgrade redis plan anymore.

**Comment by sileht** [comment]
It could be merged automatically, but I would like to be around when it reaches production :)

File: mergify_engine/__init__.py [hunk 3]
```diff
@@ -48,3 +55,64 @@ def fixup_sentry_reporting(event, hint):  # pragma: no cover
...
+        "DELETE",
+        "POST",
+        "PATCH",
+    ],
+    raise_on_status=False,
+)
+
+
+SINGLE_PULL_API_RE = re.compile(
+    f"^https://api.{config.GITHUB_DOMAIN}:443/repos/[^/]*/[^/]*/pulls/\\d+$"
```
Comment (jd) [inline mergify_engine/__init__.py]:
```suggestion
    f"^https://api.{config.GITHUB_DOMAIN}:443/repos/[^/]+/[^/]+/pulls/\\d+$"
```

File: requirements.txt [hunk 4]
```diff
@@ -17,7 +18,7 @@ Flask==1.1.1
 Flask-Cors==3.0.8
 gevent==1.4.0
 greenlet==0.4.15
-gunicorn[gevent,setproctitle]==20.0.4
+gunicorn==20.0.4
```
Comment (jd) [inline requirements.txt]: looks wrong

Comment (sileht) [reply requirements.txt]: The real requirements are in setup.cfg, requirements.txt can be generated by `tox -egenreq` and it drops extras just like dependabot does.

File: mergify_engine/__init__.py (hunk 3 above)
Comment (sileht) [reply mergify_engine/__init__.py]: done

File: mergify_engine/__init__.py [hunk 5]
```diff
@@ -48,3 +55,64 @@ def fixup_sentry_reporting(event, hint):  # pragma: no cover
...
+        "DELETE",
+        "POST",
+        "PATCH",
+    ],
+    raise_on_status=False,
+)
+
+
+SINGLE_PULL_API_RE = re.compile(
+    f"^https://api.{config.GITHUB_DOMAIN}:443/repos/[^/]+/[^/]+/pulls/\\d+$"
```
Comment (jd) [inline mergify_engine/__init__.py]: I'd be good to have a few unit tests around this one just in case…

File: mergify_engine/__init__.py [hunk 6]
```diff
@@ -48,3 +55,64 @@ def fixup_sentry_reporting(event, hint):  # pragma: no cover
...
+        self.mount(config.WEBHOOK_APP_FORWARD_URL, dashboard_adapter)
+    if config.WEBHOOK_MARKETPLACE_FORWARD_URL:
+        self.mount(config.WEBHOOK_MARKETPLACE_FORWARD_URL, dashboard_adapter)
+
+    if config.HTTP_CACHE_URL:
+        github_adapter = CustomCacheControlAdapter(
+            max_retries=RETRY, cache=RedisCache(utils.get_redis_for_http_cache())
+        )
+    else:
+        github_adapter = requests.adapters.HTTPAdapter(max_retries=RETRY)
```
Comment (jd) [inline mergify_engine/__init__.py]: this is the same than `dashboard_adapter`. Reuse it?

File: mergify_engine/__init__.py [hunk 7]
```diff
@@ -48,3 +55,64 @@ def fixup_sentry_reporting(event, hint):  # pragma: no cover
...
+
+def cached_session_init(self, *args, **kwargs):
+    real_session_init(self, *args, **kwargs)
+
+    dashboard_adapter = requests.adapters.HTTPAdapter(max_retries=RETRY)
+    self.mount(config.SUBSCRIPTION_URL, dashboard_adapter)
+    if config.WEBHOOK_APP_FORWARD_URL:
+        self.mount(config.WEBHOOK_APP_FORWARD_URL, dashboard_adapter)
+    if config.WEBHOOK_MARKETPLACE_FORWARD_URL:
+        self.mount(config.WEBHOOK_MARKETPLACE_FORWARD_URL, dashboard_adapter)
```
Comment (jd) [inline mergify_engine/__init__.py]: hm why do we need a not-cached adapter for this one?

File: mergify_engine/__init__.py (hunk 6 above)
Comment (sileht) [reply mergify_engine/__init__.py]: done

File: mergify_engine/__init__.py (hunk 7 above)
Comment (sileht) [reply mergify_engine/__init__.py]: Cached adapter works only if API server support `etag`, our dashboard api doesn'. So, it's useless.

Comment (jd) [reply mergify_engine/__init__.py]: Is it just useless, as is it has no impact, or it does not work?
I'm just wondering if it's worth having this 4 lines.

**Comment by jd** [comment]
Can we get stats about the number of hit cache/miss?

File: mergify_engine/__init__.py (hunk 7 above)
Comment (sileht) [reply mergify_engine/__init__.py]: I double checked cachecontrol, that should be safe, only 301 and 304 code have special handling, but client that doesn't send etag should never received them, so.

File: mergify_engine/__init__.py (hunk 5 above)
Comment (sileht) [reply mergify_engine/__init__.py]: done

File: mergify_engine/__init__.py (hunk 7 above)
Comment (jd) [reply mergify_engine/__init__.py]: > but client that doesn't send etag should never received them, so.

Ok, that makes sense then!

**Comment by sileht** [comment]
> Can we get stats about the number of hit cache/miss?

I will take a look at it.

**Pull request merged**